runtime/runtime/src/config.rs

```diff
@@ -10,20 +10,42 @@ use near_runtime_fees::RuntimeFeesConfig;
 // Just re-exporting RuntimeConfig for backwards compatibility.
 pub use near_runtime_configs::RuntimeConfig;
 
+use num_bigint::BigUint;
+use num_rational::Rational;
+use num_traits::cast::ToPrimitive;
+use num_traits::pow::Pow;
+use std::convert::TryFrom;
+
 /// Describes the cost of converting this transaction into a receipt.
 #[derive(Debug)]
 pub struct TransactionCost {
     /// Total amount of gas burnt for converting this transaction into a receipt.
     pub gas_burnt: Gas,
-    /// Total amount of gas used for converting this transaction into a receipt. It includes gas
-    /// that is not yet spent, e.g. prepaid gas for function calls and future execution fees.
-    pub gas_used: Gas,
+    /// The remaining amount of gas used for converting this transaction into a receipt.
+    /// It includes gas that is not yet spent, e.g. prepaid gas for function calls and
+    /// future execution fees.
+    pub gas_remaining: Gas,
+    /// The gas price at which the gas was purchased in the receipt.
+    pub receipt_gas_price: Balance,
     /// Total costs in tokens for this transaction (including all deposits).
     pub total_cost: Balance,
     /// The amount of tokens burnt by converting this transaction to a receipt.
     pub burnt_amount: Balance,
 }
 
+/// Multiplies `gas_price` by the power of `inflation_base` with exponent `inflation_exponent`.
+pub fn safe_gas_price_inflated(
+    gas_price: Balance,
+    inflation_base: Rational,
+    inflation_exponent: u8,
+) -> Result<Balance, IntegerOverflowError> {
+    let numer = BigUint::from(*inflation_base.numer() as usize).pow(inflation_exponent);
+    let denom = BigUint::from(*inflation_base.denom() as usize).pow(inflation_exponent);
+    // Rounding up
+    let inflated_gas_price: BigUint = (numer * gas_price + &denom - 1u8) / denom;
+    inflated_gas_price.to_u128().ok_or_else(|| IntegerOverflowError {})
```

Comment:
this should be capped at `max_gas_price` right?

Reply:
This will require to propagate `max_gas_price` from genesis_config to runtime through apply_state, because it's not part of `RuntimeConfig`. While it doesn't have enough benefits, since we shouldn't be able to overflow `10^22 * 1.03^64`. It's also unlikely someone will be able to ever issue a transaction that exceeds `max_gas_price`

Reply:
Yes it is unlikely that we will overflow here when calculating `inflated_gas_price`. However, if `inflated_gas_price` is more than 10^23, the calculation of balance burnt can overflow, which will crash the entire network I believe.

Reply:
No, it can't overflow, because that would mean the account paid the fees that overflow.